lib/ukrandom: Rename and refactor lib/ukswrand

Alex-deVis opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [x] Updated relevant documentation.


### Base target

 - Architectures: x86_64 and arm64
 - Platform: kvm, linuxu
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

The library depends of `HAVE_RANDOM`.
`CONFIG_LIBUKRAND_HW=y`
`CONFIG_LIBUKRAND_CHACHA=y`

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Rename `lib/ukswrand` to `lib/ukrand` to better describe its use case. It enables users to customize their entropy source, allowing a choice between hardware-generated randomness or the output of the Chacha20 PRNG.
Both random devices (`/dev/random` and `/dev/urandom`) and `getrandom()` syscall will return random numbers based on the selected source of entropy.

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+	prompt "Initial seed"
+	default LIBUKSWRAND_INITIALSEED_TIME
+
+config LIBUKSWRAND_INITIALSEED_TIME
```
Timestamp and constant seeds are not secure, so these options must be deprecated too. In fact they are even worse than not secure as they give a false sense of security. Right now the only meaningful option is processor generated randomness, so it's either that the hardware that or nothing. In the future, as we support more TRNGs we can add more here.

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+config LIBUKSWRAND_INITIALSEED_TIME
+	bool "Platform timestamp"
+
+config LIBUKSWRAND_INITIALSEED_RDRAND
```
Should be `RDSEED`, not `RDRAND`. That is because the former provides higher brute-force resistance than the DRBG-generated `RDRAND`. Saying that, the `ukarch_random` API is architecture-independent, so you don't need to specify anything here. In conclusion what I would do is:
1. Remove all "Initial Seed" options
2. Make the entire library `depends on HAVE_RANDOM`.
3. In the implementation use the `ukarch_random_seed` to generate the seed if the system is configured to use a software PRNG.

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+	prompt "Algorithm"
+	default LIBUKSWRAND_CHACHA
+
+config LIBUKSWRAND_MWC
```
Multiply-with-carry is not a CSPRNG so it should be deprecated: https://crypto.stackexchange.com/questions/10359/what-stops-the-multiply-with-carry-rng-from-being-a-cryptographically-secure-prn

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+menuconfig LIKUKRAND
+	bool "ukrand: Software random number generator"
+	default n
```
All `default n` are redundant, as this is the default option anyway.

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+menuconfig LIKUKRAND
+	bool "ukrand: Software random number generator"
```
s/Software random number generator/Random number generation/

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+	default n
+
+if LIBUKSWRAND
+
```
You can add an option for "Processor-generated randomness" and a help option that states that this is used pass-through.

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+config LIBUKSWRAND_CHACHA
+	bool "ChaCha20"
+	help
+	  Use ChaCha20 algorithm
```
Perhaps you can update this with a better description:
```
ChaCha20 is a fast CSPRNG based on Salsa20
```

michpappas commented on lib/ukrand/Makefile.uk:
```diff
@@ -0,0 +1,12 @@
+CINCLUDES-$(CONFIG_LIBUKRAND)	+= -I$(LIBUKRAND_BASE)/include
+CXXINCLUDES-$(CONFIG_LIBUKRAND) += -I$(LIBUKRAND_BASE)/include
+
+LIBUKRAND_SRCS-$(CONFIG_LIBUKSWRAND_MWC) += $(LIBUKRAND_BASE)/mwc.c
```
Please align the `+=` in these three.

michpappas reviewed (changes requested):
Hi @Alex-deVis, please see the comments below.

Alex-deVis commented on lib/ukrand/include/uk/random.h:
This is wrongly inferred as a rename. In reality, I deleted the deprecated algorithm `mwc` and added a `random.h` to expose pass-through randomness.

michpappas replied:
Okay, in that case you should use the new copyright header (and we should update the documentation):
```
/* SPDX-License-Identifier: BSD-3-Clause */
/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
 * Licensed under the BSD-3-Clause License (the "License").
 * You may not use this file except in compliance with the License.
 */
```

Alex-deVis commented on lib/ukrand/swrand.c:
```diff
@@ -32,23 +32,21 @@
-#ifdef CONFIG_LIBUKSWRAND_INITIALSEED_TIME
-	val = (__u32)ukplat_wall_clock();
-#endif
+	ret = ukarch_random_init();
```
@michpappas, if `uk_early_initcall(_uk_random_init);` is also called, will there be problems? If so, should we make hardware randomness a requirement for software randomness in `Config.uk`?

michpappas replied:
With the current implementation, it is not a problem to call `ukarch_random_init()` multiple times, as both the `arm64` and `x86_64` variants only check with the cpu whether the feature is available, HOWEVER you don't need two initcalls. What you can do instead is rename `_uk_swrand_init()` to `uk_rand_init()`. Moreover, `ukarch_random_init()` should not be in `uk_swrand_gen_seed32()`, but in `_uk_swrand_init()` (to be `uk_rand_init()`) at the very beginning. The part that generates the seed and initializes the the sw algorighm should be conditional to `CONFIG_UKRANDOM_CHACHA20`.

PS: Generally speaking randomness must be available early, as it's critical for things like ASLR, PAuth and MTE, but in practice our `inittab` does not have this kind of granularity yet. Instead early init calls are only early relatively to init calls of lower priority, yet they all happen at a single place in `ukboot`. Eventually this will change, and randomness will be initialized in early bootcode once. But for now you can leave it as is.

Alex-deVis commented on lib/ukrand/Makefile.uk:
```diff
@@ -0,0 +1,14 @@
+LIBUKRAND_SRCS-$(CONFIG_LIBUKSWRAND_CHACHA)     += $(LIBUKRAND_BASE)/chacha.c
+LIBUKRAND_SRCS-$(CONFIG_LIBUKSWRAND_DEVFS)      += $(LIBUKRAND_BASE)/dev.c
+
+# TODO: Changing -y to -$(CONFIG_LIBUKSWRAND) feels natural. How to check if it this breaks anything?
```
@michpappas, as the use of software randomness is no longer mandatory for this library, I think `getrandom()` should require `CONFIG_LIBUKSWRAND` to be selected in the config. Is my assumption correct?

michpappas replied:
Actually `getrandom()` should use whatever source of randomness is selected by the user, and the user should be always choose between software or hardware pass-through. For more details see the relevant comment in `Config.uk`. The same applies to all functions that are part of this library's public API. They should all have conditionals on whether to use software generated randomness or hardware pass-through.

Alex-deVis commented on lib/ukrand/include/sys/random.h:
```diff
@@ -37,6 +37,7 @@ extern "C" {
 #include <sys/types.h>
+#include <uk/arch/random.h>
 
 #define GRND_NONBLOCK     0x01
 #define GRND_RANDOM       0x02
```
Now that hardware randomness is available, should we also implement these flags as described in [man getrandom](https://man7.org/linux/man-pages/man2/getrandom.2.html)?

michpappas replied:
Yes, the syscall has to be reworked in general 👍🏼

Alex-deVis replied:
I think my previous suggestion is not relevant as none of these flags makes a difference for our usecase:
- `GRND_RANDOM` is used to ensure that you draw entropy from `/dev/random` instead of `/dev/urandom` i.e. the random numbers you get are better suited for cryptographic operations.
In our case, you choose your unique entropy source (either chacha CSPRNG or hardware pass-through) using `make menuconfig`, and therefore `getandom` will return random bytes from the previously selected source.
- `GRND_NONBLOCK` is used to avoid blocking if not enough entropy is available (e.g. to avoid blocking at boot time). This does not apply to us either, as we currently draw entropy from the hardware (via RDRAND / RNDR) so there is no concern about blocking.

Alex-deVis reviewed (commented):
@michpappas, I addressed all your comments. I left a few inline questions in regard to my changes.

michpappas commented on include/uk/arch/random.h:
```diff
@@ -31,6 +31,7 @@
 #ifndef __UKARCH_RANDOM_H__
 #define __UKARCH_RANDOM_H__
 
+#include <uk/arch/lcpu.h>
```
If the dependency is on arm-specific registers then you should add the include in the depending file, i.e. in `arch/arm/arm64/include/uk/asm/random.h`.

michpappas commented on lib/ukrand/include/uk/random.h:
```diff
@@ -0,0 +1,66 @@
+extern "C" {
+#endif
+
+static inline int _uk_random_init(void) {
```
Functions should not start with an underscore unless they are meant to indicate that they must not be called directly in favor of e.g. some wraping macro. So in this case the underscore is redundant.

michpappas commented on lib/ukrand/Makefile.uk:
```diff
@@ -3,6 +3,7 @@ $(eval $(call addlib_s,libukrand,$(CONFIG_LIBUKRAND)))
 CINCLUDES-$(CONFIG_LIBUKRAND)	                += -I$(LIBUKRAND_BASE)/include
 CXXINCLUDES-$(CONFIG_LIBUKRAND)                 += -I$(LIBUKRAND_BASE)/include
 
+LIBUKRAND_SRCS-$(CONFIG_LIBUKHWRAND)            += $(LIBUKRAND_BASE)/random.c
```
As a general rule:
* Public headers must be defined conditionally to the library.
* Private headers and sources must be defined unconditionally. If any of these depends on an optional feature, then they should be defined conditionally to that feature.

Example:
```
CINCLUDES-$(CONFIG_LIBUKRANDOM)   += ...
CXXINCLUDES-$(CONFIG_LIBUKRANDOM) += ... # make sure your headers contain C++ include guards
ASINCLUDES-$(CONFIG_LIBUKRANDOM)  += ... # if your headers needs to be included in asm files

LIBUKRANDOM_CINCLUDES-y += ... # doesn't make sense to define CXXINCLIDES for private headers

LIBUKRANDOM_SRCS-y += ....
LIBUKRANDOM_SRCS-$(LIBUKRANDOM_CHACHA20) += ....
```

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,56 @@
+
+if LIBUKRAND
+
+config LIBUKSWRAND
```
Given that:
1. In the future we will support multiple hardware RNGs
2. These devices will live under `drivers/` and will be selected via the corresponding `Drivers/` menu,
3. We will be allowing the user to select a single hardware source of randomness,

we should probably have one single option `Hardware Pass-through` in `libukrand` that will provide pass-through access to whichever device was selected in `Drivers/`.

Moreover, as "Software-generated randomness" may give the wrong impression (i.e purely software / not seeded by a TRNG, and thus not cryptographically secure) I think we should provide a flat set of options and make it clear that the software algorithm is seeded by the selected HW RNG.

So perhaps one single selection menu similar to the following will do:
```
Source:
* Hardware pass-through
* ChaCha20
```

michpappas replied:
The pattern for Kconfig options is `CONFIG_LIBNAME_FEATURE`, so all the `LIBUKSWRAND` options need to be renamed according to the convention.

michpappas reviewed (changes requested):
Hi @Alex-deVis please see the comments below.

Alex-deVis commented:
Hi, @michpappas, I have addressed your previous comments and left a note about `getrandom()` flags. At this point, this is ready for merge. I only managed to test this using KVM, both on X86 and Arm64.
It should be noted that the current library `ukrand` relies on the availability of CPU RNG capability. This means you cannot use `ukrand` if you are not targeting an Intel CPU with RDRAND support, or an Arm CPU with RNDR support.

mogasergiu commented on lib/ukrand/Config.uk:
```diff
@@ -0,0 +1,22 @@
+if LIBUKRAND
+
+choice
+	prompt "Source"
```
"Source"?

Alex-deVis replied:
It is used to state the source of random numbers you use: CPU generated or Chacha seeded by the CPU. Not sure what to rename this to. "Random Number Source" maybe?

Alex-deVis replied:
This refers to the source of random numbers, which is either CPU-generated or Chacha20, seeded by CPU. Not sure how to rename it. Maybe "Random Number Generator"?

mogasergiu replied:
or just "RNG Source"? Idk, but I would certainly not just leave it as "Source" :grin:

mogasergiu commented on lib/ukrand/swrand.c:
```diff
@@ -29,26 +29,20 @@
 
 __u32 uk_swrandr_gen_seed32(void)
 {
+	int ret;
```
Reverse xmas tree variable declarations, please.

mogasergiu commented on lib/ukrand/swrand.c:
```diff
@@ -29,26 +29,20 @@
-#endif
+	ret = ukarch_random_seed_u32(&val);
+	if (unlikely(ret))
+		UK_CRASH("Could not generate random seed\n");
```
I would rather not crash here and instead just output the message as a `uk_pr_warn`. Instead, I would initialize `val` at declaration time to a default seed value and just return that.

Alex-deVis replied:
It makes sense, I went with your suggestion.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -116,7 +116,19 @@ static inline __u32 _infvec_val(unsigned int c, const __u32 v[],
+__u32 uk_chacha_gen_seed32(void)
+{
+	int ret;
+	__u32 val;
```
Reverse xmas tree.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -116,7 +116,19 @@ static inline __u32 _infvec_val(unsigned int c, const __u32 v[],
+
+	ret = ukarch_random_seed_u32(&val);
+	if (unlikely(ret))
+		UK_CRASH("Could not generate random seed\n");
```
Same comment regarding crashing.

Alex-deVis replied:
Your other comment is to an outdated file, I am adding it here as well:
> I would rather not crash here and instead just output the message as a uk_pr_warn. Instead, I would initialize val at declaration time to a default seed value and just return that.

IIRC, we decided that `ukrand` should be used whenever you need cryptographically secure numbers, so it would make sense to crash if you cannot do that. If you just want random numbers and do not care if they are secure can't you use the PRNG from libc?

mogasergiu replied:
I think it should be the responsibility of the caller to know whether it's worth crashing the system or not on a failed function call :thinking:. I think the main issue here is the mixed semantics: `ukarch_random_seed_u32` seems to return the actual result through the argument and an error code through the actual return value of the functio, while `uk_chacha_gen_seed32` takes no arguments and supposedly returns the actual result through the actual return value of the function.

Therefore, I would maintain the same "signature" for the two functions. As far as I can tell from reading this function, it is supposed to be a wrapper. It does not seem to be doing anything other than call  `ukarch_random_seed_u32` itself (at least at the moment? maybe there is more planned?). Although I do not have as much context as you guys since I did not attend all the meetings. @michpappas WDYT?

Alex-deVis replied:
You are right, I did not notice that `uk_chacha_gen_seed32` was just a wrapper for `ukarch_random_seed_u32`. I removed the wrapper.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -116,7 +116,19 @@ static inline __u32 _infvec_val(unsigned int c, const __u32 v[],
+}
+
+void uk_chacha_init_r(struct uk_chacha_req *r, unsigned int seedc,
 		const __u32 seedv[])
```
Align this line to parantheses please.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -137,7 +149,23 @@ void uk_swrand_init_r(struct uk_swrand *r, unsigned int seedc,
+static int _uk_chacha_init(struct uk_init_ctx *ictx __unused)
+{
+	unsigned int i;
+	unsigned int seedc = 10;
```
Reverse xmas tree.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -137,7 +149,23 @@ void uk_swrand_init_r(struct uk_swrand *r, unsigned int seedc,
+
+	uk_chacha_init_r(&uk_chacha_def, seedc, seedv);
+
+	return seedc;
```
From what I can see `seedc` is always `10`? What is the point of this return? This is an initialization function. Either return 0 or an error.

Alex-deVis replied:
Right, I refactored this and forgot to change the return value.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +184,54 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
 }
+
+/* Uses the pre-initialized default generator  */
+/* TODO: Add assertion when we can test if we are in interrupt context */
```
Why not add it now? `UK_ASSERT(!ukplat_lcpu_irqs_disabled());`

Alex-deVis replied:
Will do that, this was taken as-is when refactoring.

mogasergiu replied:
So will you add it? :smile:

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +184,54 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+
+ssize_t uk_chacha_fill_buffer(void *buf, size_t buflen)
+{
+	size_t step, chunk_size, i;
```
s/size_t/__sz, same for above `buflen` argument. Also s/ssize_t/__ssz/. You can check `include/uk/arch/types.h` for more such types.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +184,54 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+	chunk_size = buflen % step;
+
+	for (i = 0; i < buflen - chunk_size; i += step)
+		*(__u32 *)((char *) buf + i) = uk_chacha_randr_u32();
```
No spaces between cast and variable.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +184,54 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+		memcpy(buf + i, &rd, chunk_size);
+	}
+
+	return buflen;
```
What is the point of returning the unchanged argument?

Alex-deVis replied:
At the moment the return value is useless, but it will be relevant if we implement `GRND_RANDOM` and do not use CPU-generated randomness.

mogasergiu replied:
Does this mean that you would overwrite this argument's value in that case? I assume this is for the non-blocking case where the syscall would supposedly return `EAGAIN`? Or did I misunderstand?

michpappas replied:
Re-opening, this conversation resolved without answering the question.

mogasergiu commented on lib/ukrand/dev.c:
```diff
@@ -54,7 +54,7 @@ static int dev_random_read(struct device *dev __unused, struct uio *uio,
 	count = uio->uio_iov->iov_len;
 
-	uk_swrand_fill_buffer(buf, count);
+	uk_chacha_fill_buffer(buf, count);
```
Ditto, the return value is not even checked for? :thinking:

Alex-deVis replied:
Fixed it.

michpappas replied:
Re-opening this as I don't see a return value being checked.

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,106 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2023, Alexandru-Gabriel Apostolescu <alexx.apostolescu@gmail.com>.
```
All licenses we use now are
```
/* SPDX-License-Identifier: BSD-3-Clause */
/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
 * Licensed under the BSD-3-Clause License (the "License").
 * You may not use this file except in compliance with the License.
 */
```

mogasergiu commented on lib/ukrand/include/uk/random.h:
```diff
@@ -0,0 +1,31 @@
+extern "C" {
+#endif
+
+#ifdef CONFIG_LIBUKRAND_CHACHA
```
The scheme we use now is
```C
#if CONFIG
..
#endif /* CONFIG */
```

Same for every other `#ifdef CONFIG`

mogasergiu replied:
This is marked as resolved by mistake probably :tongue:

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +184,54 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+	__u64 ret;
+
+	iflags = ukplat_lcpu_save_irqf();
+	ret = ((__u64) uk_chacha_randr_r(&uk_chacha_def)) << 32 | \
```
No space after a cast. And why is there backslash at the end? Also try to surround bit operations in parantheses.

mogasergiu commented on .github/CODEOWNERS:
```diff
@@ -65,7 +65,7 @@
 /lib/ubsan		@unikraft/maintainers-security
 /lib/uksp		@unikraft/maintainers-security
-/lib/ukswrand		@unikraft/maintainers-security
+/lib/ukrand		@unikraft/maintainers-security
```
Also mention in the commit message why you renamed it, not just that you renamed it.

mogasergiu reviewed (changes requested):
Hi @Alex-deVis thanks for you work! Had a brief look at the PR. I will return with another round after you update. Please make sure to run `./support/scripts/checkpatch.pl --strict <patches>` on the patches you submit.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -1,33 +1,7 @@
- * POSSIBILITY OF SUCH DAMAGE.
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
```
I am not sure how ok it is to change a previous license...

mogasergiu commented on lib/ukrand/dev.c:
```diff
@@ -1,33 +1,7 @@
- * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
- * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
- * POSSIBILITY OF SUCH DAMAGE.
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
Ditto

mogasergiu commented on lib/ukrand/dev.c:
```diff
@@ -64,7 +38,7 @@ static int dev_random_write(struct device *dev __unused,
 			    int flags __unused)
 {
-	return EPERM;
+	return -EPERM;
```
Why does this function return an `EPERM` in the first place? And why did you change its sign? If there is a good reason for this please do so in a separate commit. This commit supposedly only migrates a library.

Alex-deVis replied:
> Why does this function return an EPERM in the first place?

Not sure, should we change this to `return 0` or some other error?

> And why did you change its sign? 

I think it was a checkpatch warning.

mogasergiu replied:
Ok then this would mean I guess that the change is our of scope of this commit. You can choose to ignore it and undo the change, or create an additional commit giving a reason why you changed the sign, if you really see one.

Alex-deVis replied:
I reverted it at the time of the comment; I will resolve this.

mogasergiu commented on lib/ukrand/getrandom.c:
```diff
@@ -0,0 +1,18 @@
+		    void *, buf, size_t, buflen,
+		    unsigned int, flags)
+{
+	return uk_swrand_fill_buffer(buf, buflen);
```
Are you not getting an unused variable warning for `flags`? If so, consider using either `__unused` or `__maybe_unused`.

Alex-deVis replied:
I got no such warnings, I can still add any of the ones above. I'll go with `__unused` because we are not planning to use the flags any time soon.

Alex-deVis replied:
Not sure what its the proper syntax in case of `UK_SYSCALL_R_DEFINE`, could you provide an example?

mogasergiu replied:
Yeah I forgot it is actually not needed, the warning is not generated because all syscall registered arguments have that directive assigned to them. https://github.com/unikraft/unikraft/blob/staging/lib/syscall_shim/include/uk/syscall.h#L224

mogasergiu commented on lib/ukrand/include/sys/random.h:
```diff
@@ -0,0 +1,25 @@
+ * You may not use this file except in compliance with the License.
+ */
+
+#ifndef _SYS_RANDOM_H
```
Double underscores for guards: `__SYS_RANDOM_H__`... oh but this is from the old library... how did Github UI mess this up showing it as added lines :thinking:. Oh, it's because you changed the license.

mogasergiu commented on lib/ukrand/include/sys/random.h:
```diff
@@ -0,0 +1,25 @@
+#define GRND_NONBLOCK     0x01
+#define GRND_RANDOM       0x02
+
+ssize_t getrandom(void *buf, size_t buflen, unsigned int flags);
```
Why do we need this definition? Isn't `getrandom` a system call, or is it defined as per `libc`? As a kernel we should try and not directly use system calls. I understand that this is simply moving files but we should consider removing this IMO.

Alex-deVis replied:
I don't know how to answer this, so I am tagging @michpappas and @razvand.

mogasergiu replied:
You can mark this as resolved now I guess.

michpappas replied:
> I don't know how to answer this, so I am tagging @michpappas and @razvand.

This is not intended to be used by the kernel, but similarly to other syscalls, the libc wrapper is needed for natively compiled applications. The kernel API is provided in this library's header.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -109,15 +83,15 @@ static inline void _uk_iv_setup(struct uk_swrand *r, __u32 iv[2])
 }
 
 static inline __u32 _infvec_val(unsigned int c, const __u32 v[],
```
If you happen to have time, you could also add an additional commit adding `const` where needed to types as well as converting them to their non-libc variant. E.g. `ssize_t` to our `__ssz` (defined in `essentials.h`).

mogasergiu commented on lib/ukrand/include/uk/swrand.h:
```diff
@@ -0,0 +1,52 @@
+			const __u32 seedv[]);
+__u32 uk_swrand_randr_r(struct uk_swrand *r);
+
+__u32 uk_swrandr_gen_seed32(void);
```
newline here... oh wait I see now, this is an older file also shown as added lines because you changed the license. Please do no overwrite previous licenses.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -111,7 +111,26 @@ void uk_swrand_init_r(struct uk_swrand *r, unsigned int seedc,
+	uk_pr_info("Initialize random number generator...\n");
+
+	for (i = 0; i < seedc; i++) {
+		int ret = ukarch_random_seed_u32(&seedv[i]);
```
Define this above, together with the other local variables. You can ignore Codacy's warnings about reduced scopes.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -130,3 +149,53 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+	return ret;
+}
+
+ssize_t uk_chacha_fill_buffer(void *buf, size_t buflen)
```
Try to use our own defined types where possible: `__ssz`, `__sz`.

Alex-deVis replied:
This was changed in a later commit, when the function was renamed to `uk_random_fill_buffer`.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -111,7 +111,26 @@ void uk_swrand_init_r(struct uk_swrand *r, unsigned int seedc,
 }
 
-__u32 uk_swrand_randr_r(struct uk_swrand *r)
+static int _uk_chacha_init(struct uk_init_ctx *ictx __unused)
```
No underscore please in function name. As per guide, only use leading underscores if that variable is used wrapped within a macro. I know this is registered as the initialization but still :eyeglasses:

Alex-deVis replied:
Does this also apply for functions internally used by chacha: `_uk_rotl32`, `_uk_quarterround`, etc?

mogasergiu replied:
Yes, let me give you an example where this would not apply:
```C
#define __UK_LLSYSCALL_DEFINE(x, rtype, name, ename, rname, ...)	\
	long ename(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__));		\
	long rname(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__))		\
	{								\
		int _errno = errno;					\
		long ret;						\
									\
		errno = 0;						\
		ret = ename(						\
			UK_ARG_MAPx(x, UK_S_ARG_CAST_LONG, __VA_ARGS__)); \
		if (ret == -1)						\
			ret = errno ? -errno : -EFAULT;			\
		errno = _errno;						\
		return ret;						\
	}								\
	static inline rtype __##ename(UK_ARG_MAPx(x,			\
					UK_S_ARG_ACTUAL, __VA_ARGS__)); \
	long ename(UK_ARG_MAPx(x, UK_S_ARG_LONG, __VA_ARGS__))		\
	{								\
		long ret;						\
									\
		__UK_SYSCALL_PRINTD(x, rtype, ename, __VA_ARGS__);	\
		ret = (long) __##ename(					\
			UK_ARG_MAPx(x, UK_S_ARG_CAST_ACTUAL, __VA_ARGS__)); \
		return ret;						\
	}								\
	static inline rtype __##ename(UK_ARG_MAPx(x,			\
						  UK_S_ARG_ACTUAL_MAYBE_UNUSED,\
						  __VA_ARGS__))
#define _UK_LLSYSCALL_DEFINE(...) __UK_LLSYSCALL_DEFINE(__VA_ARGS__)
#define UK_LLSYSCALL_DEFINE(rtype, name, ...)				\
	_UK_LLSYSCALL_DEFINE(UK_NARGS(__VA_ARGS__),			\
			     rtype,					\
			     name,					\
			     __UK_NAME2SCALLE_FN(name),			\
			     __UK_NAME2SCALLR_FN(name),			\
			     __VA_ARGS__)
```
This is one of our macros. I think you can see what I mean. But you only moved the file, no need to also change the signatures yourself. I was just asking you to keep this in mind for the functions you yourself introduced in the PR.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -130,3 +149,53 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+
+	iflags = ukplat_lcpu_save_irqf();
+	ret = uk_chacha_randr_r(&uk_chacha_def);
+	ret = (ret << 32) | uk_chacha_randr_r(&uk_chacha_def);
```
Why not just
```C
ret = (uk_chacha_randr_r(&uk_chacha_def) << 32) | uk_chacha_randr_r(&uk_chacha_def);
```

mogasergiu commented on lib/ukrand/dev.c:
```diff
@@ -22,15 +22,18 @@
+	uio->uio_resid = 0;
 
-	uk_chacha_fill_buffer(buf, count);
+	ret = uk_random_fill_buffer(buf, count);
```
Is it likely that in a perfectly normal working scenario this function could return an error? If not, could you plaease wrap the below `if` with an `unlikely`? Or better yet, just
```C
return uk_random_fill_buffer(buf, count);
```

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,82 @@
+}
+uk_early_initcall(uk_hwrand_init, 0x0);
+
+#ifdef CONFIG_LIBUKRAND_HW
```
`#if`

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,82 @@
+}
+#endif /* CONFIG_LIBUKRAND_HW */
+
+#ifdef CONFIG_LIBUKRAND_CHACHA
```
Ditto

mogasergiu reviewed (changes requested):
Hi there :wave: ! Thanks for the updates. Please try not to overwrite already existing licenses. In the previous review round I was referring to the licenses you added on the newly added files. Other than that, I have a few more comments. Thanks for your works :+1:.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -137,7 +137,27 @@ void uk_swrand_init_r(struct uk_swrand *r, unsigned int seedc,
+	for (i = 0; i < seedc; i++) {
+		ret = ukarch_random_seed_u32(&seedv[i]);
+		if (unlikely(ret))
+			uk_pr_warn("Could not generate random seed\n");
```
I am wondering, is it really fine to only issue a warning here or would it be more sensible to also directly return the error code.

Alex-deVis replied:
Why not both? I find informative to have a warning where the issue occurs if it cannot be handled by the user. Also, I am not sure what happens if an init function fails. How should we treat this?

mogasergiu replied:
>  to also directly return the error code.

This implies both, yes :grin: 

> Also, I am not sure what happens if an init function fails.

This will be propagated to ukboot so Unikraft init will fail so all fine.

michpappas replied:
Being unable to get a seed is actually an error (and quite a critical one), so I would use `uk_pr_err()` here instead of a warning.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +176,53 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+	return ret;
+}
+
+ssize_t uk_chacha_fill_buffer(void *buf, size_t buflen)
```
__ssz and __sz

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +176,53 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+
+ssize_t uk_chacha_fill_buffer(void *buf, size_t buflen)
+{
+	size_t step, chunk_size, i;
```
__sz

mogasergiu commented on lib/ukrand/include/uk/random.h:
```diff
@@ -0,0 +1,31 @@
+
+__u32 uk_chacha_randr_u32(void);
+__u64 uk_chacha_randr_u64(void);
+ssize_t uk_chacha_fill_buffer(void *buf, size_t buflen);
```
__ssz and __sz

mogasergiu commented on arch/arm/arm64/include/uk/asm/random.h:
```diff
@@ -33,6 +33,8 @@
 
 #ifdef CONFIG_ARM64_FEAT_RNG
 
+#include <uk/arch/lcpu.h>
```
Could you squash this commit into the previous one? If this is the case then a `git bisect` landing on the previous commit would result in a build error for ARM64, correct? If that is the case, please squash this commit.

Alex-deVis replied:
Yes, that is the case, I will squash it.

michpappas replied:
This squashing was not required, I would prefer changes on arch to be separate. You can avoid breaking the build by moving it first in the commit chain. Please move this change back to a separate commit.

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,82 @@
+	chunk_size = buflen % step;
+
+	for (i = 0; i < buflen - chunk_size; i += step)
+		uk_random_u64((__u64 *)((char *)buf + i));
```
I see that you defined `uk_random_u64` as a method that returns an `int`? Are you going to check for the return value? Because otherwise all of the `uk_random_u*` methods that I can see follow this pattern and this is done to align with the arch defined ones I see. They seem to return 0 if the data in the destination is valid and 1 otherwise. Therefore, please check the return code and don't forget to use `unlikely`.

Alex-deVis replied:
>  They seem to return 0 if the data in the destination is valid and 1 otherwise.

Based on comments from `uk/arch/random.h` they return a negative value. I will return that value if anything fails.

mogasergiu replied:
Well then I believe the comments might be wrong. Even the return value despite being `int`, it is actually returned from an unsigned value and the instruction itself returning that value gives a positive result. Or am I missing something?

Alex-deVis replied:
I think you are right, the comments are wrong. I will add a commit soon to update the return value for `ukarch_random_*`,

Alex-deVis replied:
Done. I actually added it as two separate commits to better highlight changes, but we can always squash them.

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,82 @@
+
+	/* fill the remaining bytes of the buffer */
+	if (chunk_size > 0) {
+		uk_random_u64(&rd);
```
Ditto

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +176,53 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+	chunk_size = buflen % step;
+
+	for (i = 0; i < buflen - chunk_size; i += step)
+		*(__u32 *)((char *) buf + i) = uk_chacha_randr_u32();
```
No space between variable and casts.

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,82 @@
+}
+uk_early_initcall(uk_hwrand_init, 0x0);
+
+#if CONFIG_LIBUKRAND_HW
```
I think that according to your `Config.uk`, `CONFIG_LIBUKRAND_HW` and `CONFIG_LIBUKRAND_CHACHA` would be able to coexist. However this cannot be the case as you would get redeclaration errors. You could go with
```C
#if
...
#elif
#else
#error "Error about config not being selected"
#endif
```

Thus, I would suggest using `CONFIG_LIBUKRAND_CHACHA` as the default one in the configuration (`default y`) since it is a software based, architecture agnostic, solution.

Furthermore, I would add a dependency for each of the options that asks that they depend on each other not being enabled. However this is not exactly nice as selecting one means making the other not visible in the menu, so I would advise using a KConfig hint/comment here:
```
comment "Hint: Hardware randomization depends on ChaCha randomization being disabled"
	depends on CONFIG_LIBUKRAND_CHACHA

comment "Hint: ChaCha randomization depends on Hardware randomization being disabled"
	depends on CONFIG_LIBUKRAND_HW
```

However this still feels awkward. I feel like we should still be able to combine them somehow if we wish to add an additional layer of randomization. That being said, what is the reason this design decision was taken w.r.t. having the two mutually exclusive? I am sorry haven't attended much of the meetings, I do not have enough background about this. Do you happen to know @michpappas ?

Alex-deVis replied:
> I think that according to your Config.uk, CONFIG_LIBUKRAND_HW and CONFIG_LIBUKRAND_CHACHA would be able to coexist.

`CONFIG_LIBUKRAND_HW` is always used - either to provide random numbers directly or to seed CONFIG_LIBUKRAND_CHACHA.

> I would suggest using CONFIG_LIBUKRAND_CHACHA as the default one in the configuration (default y) since it is a software based, architecture agnostic, solution.

This was changed in this PR, we no longer support purely software random number generation. If the user needs that, he can use `rand_r` from libc.

mogasergiu replied:
> CONFIG_LIBUKRAND_HW is always used - either to provide random numbers directly or to seed CONFIG_LIBUKRAND_CHACHA.

Are you sure? I admit I misinterpreted it at first and did not see it's a `choice`. Being a "choice" means they are indeed mutually exclusive, right?. Where does HW seed ChaCha? From what I can see, ChaCha is purely software based in the current state of the PR.

Alex-deVis replied:
Sorry, I misread your comment. It is true that the constants `CONFIG_LIBUKRAND_HW` and `CONFIG_LIBUKRAND_CHACHA` are mutually exclusive. I will use the `if-elif-else` syntax.

> CONFIG_LIBUKRAND_HW is always used 

What I meant is that hardware randomness is always used.

> ChaCha is purely software based in the current state of the PR.

Check [`uk_chacha_init`](https://github.com/unikraft/unikraft/pull/1008/files#diff-2d44fcec5b2b3006cefe77e90a58d3ff5677c279ca6ac8284531bd8776524d25R150), it uses `ukarch_random_seed_u32` to generate the initialization vector.

mogasergiu reviewed (changes requested):
Hey, thanks for the update, we're getting there! I still have some questions left :male_detective:. Thanks!

mogasergiu commented:
An additional comment, as the Github UI shows some comments as `Outdated`. I am not sure they are. When a PR is reviewed, it is done so commit by commit. So if one commit added something and then the next commit after it updated it in a way that the very initial commit should have been done in the first place, that change should be in the previous commit instead.

Alex-deVis committed:
arch/x86_64/random.h: Fix return values

ukarch_random_u* should return either 0 for success or a negative value
for failure as per arch/random.h comments.

RDRAND and RDSEED will set the carry flag to 1 if the operation
succedded. Return 0 or -1 explicitly based on this.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
arch/arm64/random.h: Fix return values

ukarch_random_u* should return either 0 for success or a negative value
for failure as per arch/random.h comments.

RNDR and RNDRRS will set NZCV system register to 0 if the operation
succedded. Return 0 or -1 explicitly based on this.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis commented:
I covered and resolved most of the comments. I am not sure how should I handle the failed checkpatch action as it complains about the license for `swrand.c` (which was only moved and then removed in the following commit) and `random.c` whose license looks good to me.

mogasergiu commented on arch/arm/arm64/include/uk/asm/random.h:
```diff
@@ -52,6 +52,10 @@ static inline int ukarch_random_u64(__u64 *val)
 			     "	mrs	%x1, NZCV\n" /* Get result */
 			     : "=r"(*val), "=r"(res));
+
+	if (res != 0)
```
`unlikely`?

mogasergiu commented on arch/arm/arm64/include/uk/asm/random.h:
```diff
@@ -74,6 +78,10 @@ static inline int ukarch_random_seed_u64(__u64 *val)
 			     "	mrs	%x1, NZCV\n"   /* Get result */
 			     : "=r"(*val), "=r"(res));
+
+	if (res != 0)
```
Ditto

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,89 @@
+}
+uk_early_initcall(uk_hwrand_init, 0x0);
+
+#if defined CONFIG_LIBUKRAND_HW
```
Do you need the `defined`?

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,89 @@
+}
+
+#elif defined CONFIG_LIBUKRAND_CHACHA
+int uk_random_u32(__u32 *val)
```
How about we place the `#if`'s inside the functions themselves, to reduce code duplication. E.g.
```C
int uk_random_seed_u64(__u64 *val)
{
       int rc = 0;
       
#if CONFIG_LIBUKRAND_HW
	rc = ukarch_random_seed_u64(val);
#elif CONFIG_LIBUKRAND_CHACHA
       *val = uk_chacha_randr_u64();
 #else /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
 #error "Invalid RNG source"
 #endif /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
 
       return rc;
}
```

Alex-deVis replied:
Its cleaner this way. Thanks!

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,89 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
Probably 2024 now? :grin:

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -126,23 +126,45 @@ void uk_swrand_init_r(struct uk_swrand *r, unsigned int seedc,
-__u32 uk_swrand_randr_r(struct uk_swrand *r)
+static int uk_chacha_init(struct uk_init_ctx *ictx __unused)
+{
+	unsigned int seedc = 10;
```
I don't remember what the `10` was for and I can't find the comment in this PR. But can you leave a comment here maybe explaining this value? Ditto for line below

michpappas replied:
It would be preferable if this was moved to `#define` together with the explanatory comment.

mogasergiu commented on lib/ukrand/chacha.c:
```diff
@@ -156,3 +178,34 @@ __u32 uk_swrand_randr_r(struct uk_swrand *r)
+{
+	unsigned long iflags;
+	__u64 ret;
+
```
This one probably also needs a `UK_ASSERT(!ukplat_lcpu_irqs_disabled());`

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,89 @@
+		memcpy((char *)buf + i, &rd, chunk_size);
+	}
+
+	return buflen;
```
Can you also add a coment as to why we return the unmodified argument itself?

mogasergiu commented on lib/ukrand/include/sys/random.h:
```diff
@@ -32,6 +32,7 @@
 
```
Ok so yeah I don't really like this file at all. It is a copy-paste to begin with. I would like to have [musl's random.h](https://github.com/bminor/musl/blob/master/include/sys/random.h) (which is also probably the origin of this copy-paste) as a file of ours in `lib/nolibc/musl-imported/include/sys/random.h` (just like the others there) instead.

This should be, however, a more general cleanup separate commit. If you'd like you could push an additional commit into this PR deleting this file and adding the one I suggested. Otherwise we could leave it for later, no problem, won't stop me from approving this PR.

mogasergiu reviewed (changes requested):
Hi, this should be the last review :muscle:. Afterwards I will test this and see if everything is fine at runtime as well :raised_hands:. Thank you for your cooperation and your work.

mogasergiu commented:
> I covered and resolved most of the comments. I am not sure how should I handle the failed checkpatch action as it complains about the license for `swrand.c` (which was only moved and then removed in the following commit) and `random.c` whose license looks good to me.

Yeah this is part of the new checkpatch updates imported from Linux. Ignore them for now, all fine :grin::+1:

mogasergiu commented on lib/ukrand/include/uk/random.h:
```diff
@@ -0,0 +1,36 @@
+extern "C" {
+#endif
+
+#ifdef CONFIG_LIBUKRAND_CHACHA
```
`#if`, marked as resolved by mistake perhaps :).

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,103 @@
+	return rc;
+}
+
+int uk_random_u64(__u64 *val)
```
Additionally, I think it would also be good to add a `UK_ASSERT(val)` here to make sure it is not `NULL`. Ditto for `uk_random_u32`.

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,103 @@
+	return rc;
+}
+
+__ssz uk_random_fill_buffer(void *buf, __sz buflen)
```
Ditto for `buf` and `buflen`.

Alex-deVis replied:
Went with your suggestion, but maybe returning 0 instead would be more appropriate as it leaves the applications decide whether to fail at runtime or not.

mogasergiu replied:
Ok sure makes sense

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,103 @@
+
+	step = sizeof(__u64);
+	chunk_size = buflen % step;
+
```
What happens if `buflen < step` ? Wouldn't the function do nothing and falsely return it filled in `buflen` bytes? Should we add an assert?

Alex-deVis replied:
I think this case is covered, I'll leave an example to be sure we are on the same page:
```
// Assume buflen = 3
step = sizeof(__u64); // 8
chunk_size = buflen % step; // 3 % 8 = 3

// buflen = chunk_size = 3, so the for loop condition is not met (i < buflen - chunk_size)

// this will be true
if (chunk_size > 0) {
  ...
  memcpy((char *)buf + i, &rd, chunk_size); // memcpy((char *)buf + 0, &rd, 3);
  
// buflen = chunk_size = 3 as expected
return buflen;
```
Correct me if I am wrong.

mogasergiu commented on lib/ukrand/random.c:
```diff
@@ -0,0 +1,103 @@
+{
+	int rc = 0;
+
+#if CONFIG_LIBUKRAND_HW
```
We could probably make it even cleaner :neckbeard: :point_up: :nerd_face: 
```C
#if CONFIG_LIBUKRAND_HW
#define UK_RANDOM_SEED(valp, bits)			\
	({						\
		ukarch_random_u##bits((valp));		\
	})
#elif CONFIG_LIBUKRAND_CHACHA
#define UK_RANDOM_SEED(valp, bits)			\
	({						\
		*(valp) = uk_chacha_randr_u##bits();	\
		0;					\
	})
#else /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
#error "Invalid RNG source"
#endif /* !CONFIG_LIBUKRAND_HW && !CONFIG_LIBUKRAND_CHACHA */
```
and then you could use like
```C
int uk_random_u32(__u32 *val)
{
             return UK_RANDOM_SEED(val, 32);
 }
 ```

mogasergiu reviewed (changes requested):
Hey so I've cloned the code locally, seems fine. Some additional comments.

Alex-deVis commented:
@mogasergiu I added a new commit addressing the refactoring for `uk_random_u*()` functions. If everything is alright, I'll squash it.

mogasergiu commented on lib/ukrand/include/uk/swrand.h:
```diff
@@ -0,0 +1,52 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
Why did you overwrite license?

mogasergiu commented on lib/ukrand/swrand.c:
```diff
@@ -0,0 +1,63 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2023, Unikraft GmbH and The Unikraft Authors.
```
ditto

mogasergiu reviewed (approved):
Thank you for your work and patience! Make sure to squash the last commit as you have also mentioned in its commit message, I like it :grin: . I approve this PR... but as we discussed during the security meeting this would have to be rebased soon, hopefully not much. Once that is done and this is rebased on top of that (hopefully it shouldn't be much) I will approve :+1: the PR properly again.

I have noticed there are two licenses that you overwrote? They look oddly similar to the initial source files... I see that you overwrote those files entirely afterwards so that is fine, but the initial commit where you do just the renaming should maintain the initial licenses... it's just a nitpick.

Otherwise, a suggestion for future improvement would be to use intrinsics instead of raw inline assembly, e.g. `__builtin_ia32_rdrand64_step` and co for `x86_64` or `__builtin_aarch64_rndr` and co for `arm64`. But this is out of scope of this PR since you did not add them, so it's just a final thought.

With that being said, thanks again for your cooperation :muscle:!

Reviewed-by: Sergiu Moga <sergiu@unikraft.io>

michpappas commented on lib/ukrand/include/uk/random.h:
```diff
@@ -0,0 +1,31 @@
+ * You may not use this file except in compliance with the License.
+ */
+
+#ifndef __UK_RANDOM_H__
```
This is the main header exported by the library. The current (new) convention is that the library, its main header, and the functions in the API to match, so it either has to be `random` or `rand`. I would suggest that we rename the library to `libukrandom` as "random" is used in the relevant syscall as well as the exported devices (`/dev/random`, `/dev/urandom`).

michpappas commented on lib/ukrand/include/uk/random.h:
```diff
@@ -0,0 +1,31 @@
+
+#if CONFIG_LIBUKRAND_CHACHA
+
+__u32 uk_chacha_randr_u32(void);
```
Algorithm-specific calls should not be exported by the API, as a change in the algorithm will break depending code.

michpappas commented on lib/nolibc/musl-imported/include/sys/random.h:
I don't know the background behind that header's copyright, but as you mention in the commit message it looks like a copy from musl. What I would suggest is to drop the old header completely and import a new one from musl (in a separate commit) to regain its original copyright.

michpappas commented on lib/ukrand/Config.uk:
```diff
@@ -6,7 +6,12 @@ if LIBUKRAND
 choice
 	prompt "RNG Source"
-	default LIBUKRAND_CHACHA
+	default LIBUKRAND_HW
```
I think we need to rework this. Linux provides a `/dev/hwrng` for direct access to the hardware RNG. For backwards compatibility we should do the same. This means that the "RNG Source" option no longer makes sense. What we should do instead is:
* Add `/dev/hwrng` that provides pass-through access to the driver API
* Have ChaCha implement the high-level kernel API provided by this library (seeded via the RNG driver).
* Make `/dev/random` and `/dev/unrandom` use the high-level kernel API.

While we have only one software algorithm, there is no need for a "RNG source" config option, so we can remove this altogether. Moreover, the driver part will be selectable from the top-level `Drivers` Kconfig menu, and drivers will expose a single interface to this libray. Once this is merged I  will come with a follow up PR to migrate my previous work to `drivers/`. I also have a few more drivers in my TODO list.

michpappas commented on arch/arm/arm64/include/uk/asm/random.h:
```diff
@@ -52,6 +52,10 @@ static inline int ukarch_random_u64(__u64 *val)
 			     "	mrs	%x1, NZCV\n" /* Get result */
 			     : "=r"(*val), "=r"(res));
+
+	if (unlikely(res != 0))
```
If we change this to return a negative number, we may also stick to the convention and return a value from errno. Otherwise we should update the documentation instead to state "non-zero on error".

Alex-deVis replied:
I am okay with either approach and will let you decide which is more appropriate.

michpappas replied:
Okay in that case, let's return `-EIO` in all failing cases on both `arm64` and `x86_64`. Another possibility would had been to return `-EAGAIN` in cases where the retry limit is reached, but given the amount of retries we might as well consider failure as a persistent error.

michpappas commented on lib/ukrand/chacha.c:
```diff
@@ -31,65 +31,65 @@
+#include <uk/random.h>
 
-struct uk_swrand {
+struct uk_chacha_req {
```
I think an algorithm-specific naming is not needed here, as the algorithm should implement the (algorithm-agnostic) kernel API available to the rest of Unikraft. That API should also be used in the devices exported by this library. This is especially true given that on a given build we will always enable one algorithm, even if we end up providing more than one (TBD).

michpappas commented on lib/ukrand/exportsyms.uk:
```diff
@@ -1,8 +1,7 @@
-uk_swrand_randr_r
-uk_swrandr_gen_seed32
-uk_swrand_fill_buffer
+uk_chacha_randr_u32
```
As discussed below, these should not be exported by this library. I also noticed that one of the follow up commit in this PR removes them again. This should be avoided in general: A PR should present the final state to of the code to merge, not the development history.

DeliaPavel reviewed (approved):
Reviewed-by: Delia Pavel <delia_maria.pavel>

michpappas reviewed (changes requested):
@Alex-deVis I provide some more comments below. The most significant change involves the addition of `/dev/hwrng`. Regarding the initialization, in #1373 we introduce `ukplat_early_init()` for `kvm/arm64`, and the plan is to move that there. There is no x86_64 part for now.

Alex-deVis committed:
lib/ukrandom: Rename ukswrand to ukrandom

This commit renames ukswrand library to ukrandom to better describe its
use case. This library will provide cryptographically secure random
numbers, either directly from the hardware, or from a PRNG seeded with
the true random seed.

swrand.h is the main header of the library and should match its name,
therefore it is renamed to random.h. swrand.c is also renamed to match
the header.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
arm64/random.h: Include arch/lcpu.h

arm64/include/uk/asm/random.h requires ID_AA64ISAR0_EL1_RNDR_SHIFT and
ID_AA64ISAR0_EL1_RNDR_MASK which are defined in uk/arch/lcpu.h.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
lib/ukrandom: Deprecate MWC

Deprecate multiply-with-carry (MWC) PRNG because it is not
cryptographycally secure.
Remove the choice for an algorithm in Config.uk as ChaCha20 is the only
supported algorithm.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
lib/ukrandom: Seed PRNG from hardware

Remove insecure seeding options and only use hardware randomness to seed
the PRNG.

uk_swrand_init() requires ukarch_random_init() to be called beforehand,
hence they are wrapped inside ukarch_random_init().

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
lib/nolibc: Add sys/random.h

Add musl's sys/random.h implementation to nolibc and use it to replace
lib/ukrandom's sys/random.h.

Use nolibc-internal/shareddefs.h instead of types.h.
Add names for function arguments as checkpatch requires.

Source of import:
Repository: https://git.musl-libc.org/cgit/musl
Tag: v1.2.4 (f5f55d65)
Path: musl/include/sys/random.h

CheckPatch-Ignore: COMMIT_LOG_USE_LINK
CheckPatch-Ignore: SPDX_LICENSE_TAG

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
lib/ukrandom: Register /dev/hwrng

This commit introduces a new device node, /dev/hwrng, that provides
pass-through access to hardware randomness.

Hardware randomness will be available for the kernel API once
ukarch_random is implemneted as a driver. For the time being, restrict
the use of hardware randomness to /dev/hwrng.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
lib/ukrandom: Clean ChaCha20 implementation

This commit covers the esthetical issues of the chacha implementation.
It replaces size_t and ssize_t with __sz and __ssz, and it removes the
leading underscore from functions name as it does not abide to
Unikraft's current convention.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis committed:
lib/ukrandom: Add interrupt context check for ChaCha

Solve a previous TODO by checking that interruots are not disabled.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

Alex-deVis commented:
@michpappas I addressed the review comments, including renaming the library to `lib/ukrandom`. Please take another look and tell me if I missed something.

michpappas commented on lib/ukrandom/random.c:
```diff
@@ -75,18 +66,30 @@ ssize_t uk_swrand_fill_buffer(void *buf, size_t buflen)
+}
+
 uk_early_initcall(_uk_swrand_init, 0x0);
+uk_early_initcall(uk_hwrand_init, 0x0);
```
Here you have no guarantees that `uk_hwrand_init()` will be called before `uk_swrand_init()`, i.e. there is the risk that you use the RNG before you initialize it. What you can do instead is to have a single `uk_random_init()` that calls both `ukarch_random_init()` and `uk_swrand_init()` in the right order.

```
static int uk_random_init(struct uk_init_ctx *ictx __unused)
{
    int res;

    res = ukarch_random_init();
    if (unlikely(res))
        return res;

    return uk_swrand_init();
}

uk_early_initcall(uk_random_init, 0);
```

Moreover, the implementation of `uk_swrand_init()` should be in `chacha.c`, given that the seed size is specific to ChaCha. What I would to to fix this is:
* Move the implementation of `uk_swrand_init()` to `chahca.c` and make it return zero on success or negative value on error.
* In `chacha.c` rename `uk_swrand_init_r` to `chahca_init()` and make it `static`.

michpappas commented on lib/ukrandom/random.c:
```diff
@@ -32,26 +32,17 @@
+ * integers for the nonce, hence seedc is always 10.
+ * RFC7539 specifies three 32-bit integers for the nonce, but the reference
+ * implementation uses only two:
+ * http://cr.yp.to/streamciphers/timings/estreambench/submissions/salsa20/chacha8/ref/chacha.c
```
From a quick look on RFC7539 I think the reasoning of the discrepancy on the nonce size with the paper is described in Sect. 2.3:

> Note also that the original ChaCha had a 64-bit nonce and 64-bit block count.  We have modified this here to be more consistent with recommendations in [Section 3.2 of [RFC5116]](https://datatracker.ietf.org/doc/htmlrfc5116#section-3.2). This limits the use of a single (key,nonce) combination to 2^32 blocks, or 256 GB, but that is enough for most uses.

I consider reviewing ChaCha out of scope of this PR, but if you wish you can add a TODO to bring the nonce-size  in-line with the RFC.

michpappas commented on lib/ukrandom/random.c:
```diff
@@ -32,26 +32,17 @@
-	return val;
-}
+/*
+ * seedc is the seed length in 32-bit integers.
```
I wouldn't use the names of locals, i.e. `seedc` in this description. You could skip the first sentence, as the definition's name (`CHACHA_SEED_LENGTH`) makes it redundant, and rephrase the second one to "ChaCha20 requires eight 32-bit integers for the key and two 32-bit integers for the nonce, hence the seed length is 10."

michpappas commented on lib/nolibc/musl-imported/include/sys/random.h:
```diff
@@ -0,0 +1,42 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
```
The original file https://git.musl-libc.org/cgit/musl/tree/include/sys/random.h doesn't include that copyright header, and I think you don't need to add it as it is covered by https://github.com/unikraft/unikraft/blob/staging/lib/nolibc/musl-imported/COPYRIGHT.

Also in the commit message please provide clear traceability information, similarly to:
https://github.com/unikraft/unikraft/pull/1244/commits/13129a2c3a3a82bd4340e9b79e1697cda33b3e5a and document any chances you have made on top of the import (see comment below on `nolibc/shareddefs.h`).

michpappas commented on lib/nolibc/musl-imported/include/sys/random.h:
```diff
@@ -0,0 +1,42 @@
+extern "C" {
+#endif
+
+#include <sys/types.h>
```
Here you need to keep the definitions of `__NEED_size_t` and `__NEED_ssize_t` and include `<nolibc/shareddefs.h>` instead of `sys/types.h`.

michpappas commented on lib/ukrandom/dev.c:
```diff
@@ -44,20 +44,21 @@
 {
 	size_t count;
 	char *buf;
+	int ret;
```
`int ret` looks like an error code. Since `uk_swrand_fill_buffer()` cannot fail, I would rename this to `bytes` or similar. Also the type should be `__ssz`.

michpappas commented on lib/ukrandom/include/uk/random.h:
```diff
@@ -67,6 +67,7 @@ static inline __u32 uk_swrand_randr(void)
 }
 
 ssize_t uk_swrand_fill_buffer(void *buf, size_t buflen);
+ssize_t uk_random_fill_buffer(void *buf, __sz buflen);
```
Maybe you can keep this one private to the library for now. I will define a separate API that provides direct access to the hwrng when I move `ukarch_random` to drivers. Of course this means that until that happens you will need to add a declaration in `dev.c`.

Remember also to remove the entry from `exportsyms.uk`.

michpappas commented on lib/ukrandom/dev.c:
```diff
@@ -87,12 +103,27 @@ static struct driver drv_urandom = {
 
-	uk_pr_info("Register '%s' and '%s' to devfs\n",
-		   DEV_URANDOM_NAME, DEV_RANDOM_NAME);
+	uk_pr_info("Register '%s', %s, and '%s' to devfs\n",
```
You also need to update the message in `Config.uk`:
```
bool "Register random, urandom, and hwrng devices to devfs"
```

michpappas reviewed (changes requested):
@Alex-deVis many thanks for the updates. I think after this round we can merge.

Besides the inline comments you'll find below, the library's API exported in `libukrandom/include/uk/random.h` should match the library's name, i.e. be renamed from `uk_swrandom_*` to `uk_random_*`, and be documented with doxygen.

Last but not least, here are some typos I spotted in the commit messages:

**lib/ukrandom: Rename ukswrand to ukrandom**

"cryptographycall secure" -> "cryptographically secure"

**lib/ukrandom: Seed PRNG from hardware**

"Remove unsecure" -> "Remove insecure"

"randomness to see" -> "randomness to seed"
    
**lib/ukrandom: Register /dev/hwrng**

"which exposes true randomness seeded by the hardware" ->  "which provides pass-through access to the hwrng"

**lib/ukrandom: Clean ChaCha20 implementation**

"It replaces size_t and ssise_t with __sz and __sz" -> "It replaces size_t and ssize_t with __sz and __ssz"

Alex-deVis committed:
lib/ukrandom: Clean the public API

This commit addresses the public API by renaming the main randomness
function uk_swrand_fill_buffer() to uk_random_fill_buffer(), to abide
Unikraft's naming convention.

As a result, the private hwrng function fill_buffer() is renamed to
uk_random_hwrng_fill_buffer().

uk_random_fill_buffer() is the result of a PRNG and it never fails.
Make it void to reflect this.

Signed-off-by: Alex Apostolescu <alexx.apostolescu@gmail.com>

michpappas commented on lib/ukrandom/dev.c:
```diff
@@ -44,18 +45,46 @@
-	uk_swrand_fill_buffer(buf, count);
-
+	uk_random_fill_buffer(buf, count);
+	
```
Stray tabs here. This should had been caught by the checkpatch check, yet it did not. It seems to be related to govctl as the issue is properly reported if I run `checkpatch.uk` against that commit from the command-line.

CC: @craciunoiuc

michpappas reviewed (approved):
Looks good. Thanks again @Alex-deVis for going through the lengthy processes to get that merged.

Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
@Alex-deVis before we merge could you please add an error message in case the hwrng cannot be initialized / is not supported:
```diff
diff --git a/lib/ukrandom/random.c b/lib/ukrandom/random.c
index 5bf82729f..a200b8f59 100644
--- a/lib/ukrandom/random.c
+++ b/lib/ukrandom/random.c
@@ -61,8 +61,10 @@ static int uk_random_init(struct uk_init_ctx *ictx __unused)
        int res;
 
        res = ukarch_random_init();
-       if (unlikely(res))
+       if (unlikely(res)) {
+               uk_pr_err("Could not initialize the HWRNG (%d)\n", res);
                return res;
+       }
 
        return uk_swrand_init();
 }
```

razvand reviewed (approved):
Approved-by: Razvan Deaconescu <razvand@unikraft.io>
